p2p/cuprate-p2p/src/connection_maintainer.rs

```diff
@@ -0,0 +1,290 @@
+//! Outbound Connection Maintainer.
+//!
+//! This module handles maintaining the number of outbound connections defined in the [`P2PConfig`].
+//! It also handles making extra connections when the peer set is under load or when we need data that
+//! no connected peer has.
+use std::sync::Arc;
+
+use rand::{distributions::Bernoulli, prelude::*};
+use tokio::{
+    sync::{mpsc, OwnedSemaphorePermit, Semaphore},
+    task::JoinSet,
+    time::{sleep, timeout},
+};
+use tower::{Service, ServiceExt};
+use tracing::instrument;
+
+use monero_p2p::{
+    client::{Client, ConnectRequest, HandshakeError},
+    services::{AddressBookRequest, AddressBookResponse},
+    AddressBook, NetworkZone,
+};
+
+use crate::{
+    client_pool::ClientPool,
+    config::P2PConfig,
+    constants::{HANDSHAKE_TIMEOUT, MAX_SEED_CONNECTIONS, OUTBOUND_CONNECTION_ATTEMPT_TIMEOUT},
+};
+
+enum OutboundConnectorError {
+    MaxConnections,
+    FailedToConnectToSeeds,
+    NoAvailablePeers,
+}
+
+/// A request from the peer set to make an outbound connection.
+///
+/// This will only be sent when the peer set is under load from the rest of Cuprate or the peer
+/// set needs specific data that none of the currently connected peers have.
+pub struct MakeConnectionRequest {
+    /// The block needed that no connected peers have due to pruning.
+    block_needed: Option<u64>,
+}
+
+/// The outbound connection count keeper.
+///
+/// This handles maintaining a minimum number of connections and making extra connections when needed, upto a maximum.
+pub struct OutboundConnectionKeeper<N: NetworkZone, A, C> {
+    /// The pool of currently connected peers.
+    pub client_pool: Arc<ClientPool<N>>,
+    /// The channel that tells us to make new _extra_ outbound connections.
+    pub make_connection_rx: mpsc::Receiver<MakeConnectionRequest>,
+    /// The address book service
+    pub address_book_svc: A,
+    /// The service to connect to a specific peer.
+    pub connector_svc: C,
+    /// A semaphore to keep the amount of outbound peers constant.
+    pub outbound_semaphore: Arc<Semaphore>,
+    /// The amount of peers we connected to because we needed more peers. If the `outbound_semaphore`
+    /// is full, and we need to connect to more peers for blocks or because not enough peers are ready
+    /// we add a permit to the semaphore and keep track here, upto a value in config.
+    pub extra_peers: usize,
+    /// The p2p config.
+    pub config: P2PConfig,
+    /// The [`Bernoulli`] distribution, when sampled will return true if we should connect to a gray peer or
+    /// false if we should connect to a white peer.
+    ///
+    /// This is weighted to the percentage given in `config`.
+    pub peer_type_gen: Bernoulli,
+}
+
+impl<N: NetworkZone, A, C> OutboundConnectionKeeper<N, A, C>
+where
+    A: AddressBook<N>,
+    C: Service<ConnectRequest<N>, Response = Client<N>, Error = HandshakeError>,
+    C::Future: Send + 'static,
+{
```

Comment:
nit
```suggestion
impl<N, A, C> OutboundConnectionKeeper<N, A, C>
where
    A: AddressBook<N>,
    C: Service<ConnectRequest<N>, Response = Client<N>, Error = HandshakeError>,
    C::Future: Send + 'static,
    N: NetworkZone,
{
```

Reply:
Btw this is because `cargo doc` formats it as-is:

![image](https://github.com/Cuprate/cuprate/assets/101352116/162689f9-d577-41a3-a6af-33db31662633)